cilium/bpf_metadata.cc

```diff
@@ -208,6 +208,64 @@ uint32_t Config::resolveSourceIdentity(const PolicyInstanceConstSharedPtr policy
   return source_identity;
 }
 
+SourceIPs
+Config::getL7LBEastWestSourceIPs(const PolicyInstanceConstSharedPtr policy,
+                                 const Network::Address::Ip* sip,
+                                 const Network::Address::InstanceConstSharedPtr sourceAddress) {
+  struct SourceIPs sourceIPs;
+
+  const auto& ips = policy->getEndpointIPs();
+  switch (sip->version()) {
+
+    // Keep the original source address for the matching IP version, create a new source IP for
+    // the other version (with the same source port number) in case an upstream of a different
+    // IP version is chosen.
+  case Network::Address::IpVersion::v4:
+    sourceIPs.ipv4 = sourceAddress;
+    if (ips.ipv6_) {
+      sockaddr_in6 sa6 = *reinterpret_cast<const sockaddr_in6*>(ips.ipv6_->sockAddr());
+      sa6.sin6_port = htons(sip->port());
+      sourceIPs.ipv6 = std::make_shared<Network::Address::Ipv6Instance>(sa6);
+    }
+    break;
+  case Network::Address::IpVersion::v6:
+    sourceIPs.ipv6 = sourceAddress;
+    if (ips.ipv4_) {
+      sockaddr_in sa4 = *reinterpret_cast<const sockaddr_in*>(ips.ipv4_->sockAddr());
+      sa4.sin_port = htons(sip->port());
+      sourceIPs.ipv4 = std::make_shared<Network::Address::Ipv4Instance>(&sa4);
+    }
+    break;
+  }
+
+  return sourceIPs;
+}
+
+const Network::Address::Ip*
+Config::getL7LBNorthSouthSourceIP(const Network::Address::Ip* sip,
+                                  Network::Address::InstanceConstSharedPtr ipv4SourceAddress,
+                                  Network::Address::InstanceConstSharedPtr ipv6SourceAddress) {
+  // North/south L7 LB, assume the source security identity of the configured source addresses,
+  // if any and policy for this identity exists.
+  const Network::Address::Ip* ip = nullptr;
+
+  // Pick the local source address of the same family as the incoming connection
+  switch (sip->version()) {
+  case Network::Address::IpVersion::v4:
+    if (ipv4SourceAddress) {
+      ip = ipv4SourceAddress->ip();
+    }
+    break;
+  case Network::Address::IpVersion::v6:
+    if (ipv6SourceAddress) {
+      ip = ipv6SourceAddress->ip();
+    }
+    break;
+  }
+
+  return ip;
+}
+
```

Comment:
This could be more generic `selectIPVersion()`, if the first argument would be `Network::Address::IpVersion`, i.e.:
```
const Network::Address::Ip*
Config::selectIPVersion(Network::Address::IpVersion version,
                        Network::Address::InstanceConstSharedPtr ipv4Address,
                        Network::Address::InstanceConstSharedPtr ipv6Address) {
  ...
```

Reply:
good point. i changed it accordingly.